sled-agent/config-reconciler/src/reconciler_task/external_disks.rs

```diff
@@ -0,0 +1,1020 @@
+                    InlineErrorChain::new(&err),
+                );
+                ExternalDiskState::failed(config, err)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ExternalDiskState {
+    config: OmicronPhysicalDiskConfig,
+    state: DiskState,
+}
+
+impl ExternalDiskState {
+    fn managed(config: OmicronPhysicalDiskConfig, disk: Disk) -> Self {
+        Self { config, state: DiskState::Managed(disk) }
+    }
+
+    fn failed(
+        config: OmicronPhysicalDiskConfig,
+        err: DiskManagementError,
+    ) -> Self {
+        Self { config, state: DiskState::FailedToManage(err) }
+    }
+}
+
+impl IdMappable for ExternalDiskState {
+    type Id = PhysicalDiskUuid;
+
+    fn id(&self) -> Self::Id {
+        self.config.id
+    }
+}
+
+#[derive(Debug)]
+enum DiskState {
+    Managed(Disk),
+    FailedToManage(DiskManagementError),
+}
+
+/// Helper to allow unit tests to run without interacting with the real [`Disk`]
+/// implementation. In production, the only implementor of this trait is
+/// [`RealDiskAdopter`].
+trait DiskAdopter {
+    fn adopt_disk(
+        &mut self,
+        raw_disk: RawDisk,
+        mount_config: &MountConfig,
+        pool_id: ZpoolUuid,
+        log: &Logger,
+    ) -> impl Future<Output = Result<Disk, DiskManagementError>> + Send;
+}
+
+struct RealDiskAdopter<'a> {
+    key_requester: &'a StorageKeyRequester,
+}
+
+impl DiskAdopter for RealDiskAdopter<'_> {
+    async fn adopt_disk(
+        &mut self,
+        raw_disk: RawDisk,
+        mount_config: &MountConfig,
+        pool_id: ZpoolUuid,
+        log: &Logger,
+    ) -> Result<Disk, DiskManagementError> {
+        Disk::new(
+            log,
+            mount_config,
+            raw_disk,
+            Some(pool_id),
+            Some(self.key_requester),
+        )
+        .await
+        .map_err(|err| {
+            let err_string = InlineErrorChain::new(&err).to_string();
+            match err {
+                // We pick this error out and identify it separately because
+                // it may be transient, and should sometimes be handled with
+                // a retry.
+                DiskError::Dataset(DatasetError::KeyManager(_)) => {
+                    DiskManagementError::KeyManager(err_string)
+                }
+                _ => DiskManagementError::Other(err_string),
+            }
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use assert_matches::assert_matches;
+    use illumos_utils::zpool::ZpoolName;
+    use omicron_common::disk::DiskIdentity;
+    use omicron_test_utils::dev;
+    use omicron_uuid_kinds::ZpoolUuid;
+    use sled_hardware::DiskFirmware;
+    use sled_hardware::DiskPaths;
+    use sled_hardware::PooledDisk;
+    use sled_hardware::UnparsedDisk;
+    use std::collections::BTreeMap;
+    use test_strategy::proptest;
+
+    #[derive(Debug, Default)]
+    struct TestDiskAdopter {
+        requests: Vec<RawDisk>,
```

Comment:
AFAICT, wrapping this in an Arc<Mutex<..>>> should enable us to operate on `&self` instead of `&mut self` within `adopt_disk`